File: StockEasily-UI/src/dto/request/update-article-request-dto.ts
import type {CategoryRequestDto} from '../category-request-dto';
import type {ValidatableArticle} from '../create-article-request-dto';
import type {PropertyRequestDto} from '../property-request-dto';

export class UpdateArticleRequestDto {
    name = '';
    category: CategoryRequestDto = {name: ''};
    properties: PropertyRequestDto[] = [];
    quantity = 1;
    image?: string;

    public constructor(validatable: ValidatableArticle) {
        this.name = validatable.name.value;
        this.image = validatable.image.value;
        this.category = {name: validatable.category.value!};
        this.properties = validatable.properties.map(p => p.value!);
        this.quantity = validatable.quantity.value;
    }
}

File: StockEasily-API/src/main/java/de/throsenheim/unlimited/stockeasilyapi/dto/request/UpdateArticleRequestDto.java
package de.throsenheim.unlimited.stockeasilyapi.dto.request;

import io.swagger.annotations.ApiModelProperty;

import javax.validation.Valid;
import javax.validation.constraints.PositiveOrZero;
import javax.validation.constraints.Size;
import java.util.List;

public class UpdateArticleRequestDto {

    @ApiModelProperty(notes = "Article category", example = "Cables", required = true)
    @Valid
    private CategoryRequestDto category;

    @ApiModelProperty(notes = "Article name", example = "HDX42 Widescreen Monitor", required = true)
    @Size(min = 1, max = 30, message = "Article name must be between 1 and 30 characters")
    private String name;

    @Valid
    private List<PropertyRequestDto> properties;

    // Why zero as minimum? It might be used for memorable value
    // which is a common thing in terms of bookkeeping
    @ApiModelProperty(notes = "Article quantity", example = "2")
    @PositiveOrZero(message = "Quantity must greater than or equal to 0")
    private int quantity = -1;

    private String image;

    public CategoryRequestDto getCategory() {
        return category;
    }

    public String getImage() {
        return image;
    }

    public String getName() {
        return name;
    }
}